File: android/src/main/java/com/sip/flutter/sip_sdk_flutter/sip/SIPManage.java
package com.sip.flutter.sip_sdk_flutter.sip;

import android.content.Context;
import android.content.Intent;
import android.os.Handler;
import android.os.Looper;

import com.sip.flutter.sip_sdk_flutter.SipSdkFlutterPlugin;
import com.sip.flutter.sip_sdk_flutter.activity.CallActivity;
import com.sip.flutter.sip_sdk_flutter.activity.entity.CMSIPSDKCallParam;
import com.sip.sdk.SIPSDK;
import com.sip.sdk.entity.SIPSDKCallParam;
import com.sip.sdk.entity.SIPSDKConfig;
import com.sip.sdk.entity.SIPSDKDtmfInfoParam;
import com.sip.sdk.entity.SIPSDKLocalConfig;
import com.sip.sdk.entity.SIPSDKMediaConfig;
import com.sip.sdk.entity.SIPSDKMessageParam;
import com.sip.sdk.entity.SIPSDKRegistrarConfig;
import com.sip.sdk.i.SIPSDKListener;

import java.util.HashMap;
import java.util.Map;

import io.flutter.Log;

public class SIPManage implements SIPSDKListener.InitCompletedListener,
        SIPSDKListener.RegistryStateListener,
        SIPSDKListener.DtmfInfoListener,
        SIPSDKListener.MessageListener,
        SIPSDKListener.MessageStateListener,
        SIPSDKListener.IncomingCallListener,
        SIPSDKListener.CallStateListener,
        SIPSDKListener.ExpireWarningCallbackListener {
    private final Handler handler = new Handler(Looper.getMainLooper());
    private Context context;

    private static class Instance {
        private static final SIPManage instance = new SIPManage();
    }

    public static SIPManage instance() {
        return Instance.instance;
    }

    public void init(Context context,
                     String baseUrl,
                     String clientId,
                     String clientSecret,
                     SIPSDKConfig config,
                     SIPSDKMediaConfig mediaConfig) {
        this.context = context;
        SIPSDK.addListener(this);
        SIPSDK.init(context, baseUrl, clientId, clientSecret, config, mediaConfig);
    }

    @Override
    public void onExpireWarning(long expireTime, long currentTime) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("expireTime", expireTime);
        payload.put("currentTime", currentTime);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onExpireWarning", payload);
        });
    }

    @Override
    public void onRegistryState(int code) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("state", code);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onRegistrarState", payload);
        });
    }

    @Override
    public void onInitCompleted(int state, String msg) {
        Map<String, Object> map = new HashMap<>();
        map.put("state", state);
        map.put("message", msg);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onInitCompleted", map);
        });
    }

    @Override
    public void onIncomingCall(SIPSDKCallParam param) {
        Intent intent = new Intent(this.context, CallActivity.class);
        CMSIPSDKCallParam callParam = new CMSIPSDKCallParam(0, param);
        intent.putExtra("callParam", callParam);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        this.context.startActivity(intent);

        Map<String, Object> payload = new HashMap<>();
        payload.put("callType", callParam.callType);
        payload.put("username", callParam.username);
        payload.put("remoteIp", callParam.remoteIp);
        payload.put("headers", param.headers);
        payload.put("callUUID", String.valueOf(callParam.callUuid));
        payload.put("transmitVideo", callParam.transmitVideo);
        payload.put("transmitSound", callParam.transmitSound);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onIncomingCall", payload);
        });
    }

    @Override
    public void onCallState(long callUuid, int state) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("state", state);
        payload.put("callUUID", String.valueOf(callUuid));
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onCallState", payload);
        });
    }

    @Override
    public void onDtmfInfo(SIPSDKDtmfInfoParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("callUUID", String.valueOf(param.callUuid));
        payload.put("dtmfInfoType", param.dtmfInfoType);
        payload.put("contentType", param.contentType);
        payload.put("content", param.content);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onDtmfInfo", payload);
        });
    }

    @Override
    public void onMessage(SIPSDKMessageParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("messageType", param.messageType);
        payload.put("username", param.username);
        payload.put("remoteIp", param.remoteIp);
        payload.put("content", param.content);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onMessage", payload);
        });
    }

    @Override
    public void onMessageState(int state, SIPSDKMessageParam param) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("state", state);
        Map<String, Object> message = new HashMap<>();
        message.put("messageType", param.messageType);
        message.put("username", param.username);
        message.put("remoteIp", param.remoteIp);
        message.put("content", param.content);
        payload.put("message", message);
        handler.post(() -> {
            SipSdkFlutterPlugin.channel.invokeMethod("onMessageState", payload);
        });
    }
}

File: example/lib/sip_manage.dart
import 'package:flutter/cupertino.dart';
import 'package:sip_sdk_flutter/entitys/sip_sdk_call_param.dart';
import 'package:sip_sdk_flutter/entitys/sip_sdk_config.dart';
import 'package:sip_sdk_flutter/entitys/sip_sdk_dtmf_info.dart';
import 'package:sip_sdk_flutter/entitys/sip_sdk_local_config.dart';
import 'package:sip_sdk_flutter/entitys/sip_sdk_media_config.dart';
import 'package:sip_sdk_flutter/entitys/sip_sdk_message.dart';
import 'package:sip_sdk_flutter/entitys/sip_sdk_registrar_config.dart';
import 'package:sip_sdk_flutter/entitys/sip_sdk_stun_config.dart';
import 'package:sip_sdk_flutter/entitys/sip_sdk_turn_config.dart';
import 'package:sip_sdk_flutter/sip_sdk_callbacks.dart';
import 'package:sip_sdk_flutter/sip_sdk_flutter.dart';

class SIPManage implements SIPSDKCallbacks {
  // 私有构造函数
  SIPManage._internal();

  // 静态私有实例
  static final SIPManage _instance = SIPManage._internal();

  // 工厂构造函数返回单例
  factory SIPManage() => _instance;

  static late SipSdkFlutter _sipSdkFlutterPlugin;

  static void initialize() {
    //初始化插件
    _sipSdkFlutterPlugin = SipSdkFlutter();
    //设置回调
    _sipSdkFlutterPlugin.setupCallbacks(_instance);
    //初始化SDK
    initSDK();
  }

  // 你可以继续在这里写你自己的方法：
  static void initSDK() {
    final config = SIPSDKConfig(
      baseUrl: "https://120.79.7.237",
      clientId: "1379018005584941056",
      clientSecret: "7489ed9e086e12ab45688c0caf4a7d2b",
      userAgent: 'JHCloud-flutter-1.0',
      mediaConfig: SIPSDKMediaConfig(),
      stunConfig: STUNConfig(
        servers: ["120.79.7.237:3478"],
        enableIPv6: false,
      ),
    );
    _sipSdkFlutterPlugin.initSDK(config);
  }

  void registrar() {
    final config = SIPSDKRegistrarConfig(
      domain: "test.com",
      username: "RM-1-1-1-1-1-1",
      password: "123456",
      transport: "tcp",
      serverAddr: "43.160.204.96",
      serverPort: 5060,
      proxy: "43.160.204.96",
      proxyPort: 5060,
      enableStreamControl: false,
      streamElapsed: 0,
      startKeyframeCount: 120,
      startKeyframeInterval: 1000,
      turnConfig: SIPSDKTURNConfig(
        enable: true,
        server: "120.79.7.237:3478",
        realm: "120.79.7.237",
        username: "test",
        password: "test",
      ),
      localConfig: SIPSDKLocalConfig(
        username: "RM-1-1-1-1-1-1",
        enableStreamControl: false,
        streamElapsed: 0,
        startKeyframeCount: 120,
        startKeyframeInterval: 1000,
      ),
    );
    _sipSdkFlutterPlugin.registrar(config);
  }

  void unRegistrar() {
    _sipSdkFlutterPlugin.unRegistrar();
  }

  void call(String username, Map<String, String> headers) {
    _sipSdkFlutterPlugin.call(username, headers);
  }

  void callIP(String ip, Map<String, String> headers) {
    _sipSdkFlutterPlugin.callIP(ip, headers);
  }

  // 通常情况不用调用接听，因为被叫界面是原生代码
  void answer(int code, String callUUID) {
    _sipSdkFlutterPlugin.answer(code, callUUID);
  }

  // 通常情况不用调用发送Dtmf Info，因为被叫界面是原生代码
  void sendDtmfInfo(int type, String content, String callUUID) {
    _sipSdkFlutterPlugin.sendDtmfInfo(type, content, callUUID);
  }

  // 通常情况不用调用挂断，因为被叫界面是原生代码
  void hangup(int code, String callUUID) {
    _sipSdkFlutterPlugin.hangup(code, callUUID);
  }

  void sendMessage(String username, String content) {
    _sipSdkFlutterPlugin.sendMessage(username, content);
  }

  void sendMessageIP(String username, String content) {
    _sipSdkFlutterPlugin.sendMessageIP(username, content);
  }

  // 通常情况不用调用dump，这个主要用于调试
  void dump() {
    _sipSdkFlutterPlugin.dump();
  }

  @override
  void onInitCompleted(int state, String message) {
    debugPrint("onInitCompleted: $state    $message");
    //初始化成功，注册
    if (state == 0) {
      registrar();
    }
  }

  @override
  void onStopCompleted() {
    debugPrint("onStopCompleted");
  }

  @override
  void onRegistrarState(int state) {
    debugPrint("onRegistrarState: $state");
  }

  @override
  void onDtmfInfo(SIPSDKDtmfInfo dtmfInfo) {
    debugPrint("onDtmfInfo: ${dtmfInfo.toString()}");
  }

  @override
  void onMessage(SIPSDKMessage message) {
    debugPrint("onDtmfInfo: ${message.toString()}");
  }

  @override
  void onMessageState(int state, SIPSDKMessage message) {
    debugPrint("onDtmfInfo: $state:${message.toString()}");
  }

  @override
  void onCallState(String callUUID, int state) {
    debugPrint("onCallState: $callUUID:$state");
  }

  @override
  void onIncomingCall(SIPSDKCallParam callParam) {
    debugPrint("onIncomingCall: ${callParam.toString()}");
  }

  @override
  void onExpireWarning(DateTime expireTime, DateTime currentTime) {
    final expireStr = expireTime.toLocal().toString();
    final currentStr = currentTime.toLocal().toString();
    debugPrint(
      'License will expire at: $expireStr, current time is: $currentStr',
    );
  }
}
